### PR TITLE
feat(utils): add file-system helper

### PR DESCRIPTION
**Pull request by BioPhoton**
This PR is a result of the plugin docs and examples we are working on.
The added helper will reduce the code drastically in both docs and example code.

This PR includes:
- merging files and tests
- add 2 helper to `file-system.ts`
- introduce `formatting.ts`

### CODE REVIEW

File: packages/utils/src/lib/file-system.ts [hunk 1]
```diff
@@ -87,3 +88,42 @@ export async function importEsmModule<T = unknown>(
...
+
+export async function crawlFileSystem<T>(options: {
+  directory: string;
+  pattern?: string | RegExp;
+  fileTransform?: (filePath: string) => Promise<T> | T;
+}): Promise<T[]> {
+  const {
+    directory,
+    pattern,
+    fileTransform = (filePath: string) => filePath as unknown as T,
```
Comment (matejchalk) [inline packages/utils/src/lib/file-system.ts:100]: Can be simplified:

```suggestion
    fileTransform = (filePath: string) => filePath as T,
```

File: packages/utils/src/lib/file-system.unit.test.ts [hunk 2]
```diff
@@ -130,3 +132,86 @@ describe('importEsmModule', () => {
...
+        fileTransform: () => Promise.resolve('42'),
+      }),
+    ).resolves.toEqual([
+      expect.stringContaining('42'),
+      expect.stringContaining('42'),
+    ]);
+  });
+});
+
+describe('findLineNumber', () => {
```
Comment (matejchalk) [inline packages/utils/src/lib/file-system.unit.test.ts:200]:
```suggestion
describe('findLineNumberInText', () => {
```

**Review by matejchalk — CHANGES_REQUESTED** [review]

File: packages/utils/src/lib/transformation.ts [hunk 3]
```diff
@@ -55,3 +37,15 @@ export function deepClone<T>(obj: T): T {
...
+
+export function factorOf<T>(items: T[], filterFn: (i: T) => boolean): number {
+  const itemCount = items.length;
+  let filterCount = items.filter(filterFn).length;
+  if (itemCount < filterCount) {
+    throw new Error(
+      `Issues: ${itemCount} cannot be less than errors ${filterCount}`,
+    );
+  }
+  filterCount = Math.max(filterCount, 0);
```
[the remark this answers is not in the document]
Comment (BioPhoton) [reply packages/utils/src/lib/transformation.ts:49]: Thx for the catch!!

File: packages/utils/src/lib/transformation.ts [hunk 4]
```diff
@@ -55,3 +37,15 @@ export function deepClone<T>(obj: T): T {
...
+  return filterCount > 0 ? Math.abs((itemCount - filterCount) / itemCount) : 1;
```
[the remark this answers is not in the document]
Comment (BioPhoton) [reply packages/utils/src/lib/transformation.ts:50]: yes. leftover from the refactoring i gess

File: packages/utils/src/lib/file-system.ts [hunk 5]
```diff
@@ -87,3 +88,42 @@ export async function importEsmModule<T = unknown>(
...
+    const filePath = join(directory, file);
+    const stats = await stat(filePath);
+
+    if (stats.isDirectory()) {
+      return crawlFileSystem({ directory: filePath, pattern, fileTransform });
+    } else if (stats.isFile() && (!pattern || new RegExp(pattern).test(file))) {
+      return fileTransform(filePath);
+    } else {
+      return [];
+    }
```
Comment (matejchalk) [inline packages/utils/src/lib/file-system.ts:114]: Bit of a nitpick, but when you `return` in the `if`-block you don't need the `else` branch - helps flatten the code a bit.

```suggestion
    if (stats.isDirectory()) {
      return crawlFileSystem({ directory: filePath, pattern, fileTransform });
    }
    if (stats.isFile() && (!pattern || new RegExp(pattern).test(file))) {
      return fileTransform(filePath);
    }
    return [];
```

**Review by matejchalk — APPROVED** [review]

**Pull request merged**